File: seq-library/APTw_3T_000_2uT_1block_2s_braintumor/EVAL_APTw_3T_000_2uT_1block_2s_braintumor.m
%% EVAL APTw_3T_000_2uT_1block_2s_braintumor
% minimalistic data evaluation for data simulated or acquired with
% APTw_3T_000_2uT_1block_2s_braintumor.seq
%
% Moritz Zaiss 2023


%% PULSEQ-CEST SIMULATION
%=================================================================================================================================

%% 1)  read in associated seq file
cd('W:\radiologie\data\MR-Physik\Mitarbeiter\Schuere\pulseq-cest-library\seq-library\APTw_3T_000_2uT_1block_2s_braintumor')
seq = SequenceSBB(getScannerLimits());
gamma_hz  = seq.sys.gamma*1e-6;                  % for H [Hz/uT]
seq.read('APTw_3T_000_2uT_1block_2s_braintumor.seq');
defs.offsets_ppm   = seq.definitions('offsets_ppm');
Nmeas=numel(defs.offsets_ppm);


%% 2)  read in data from simulation and Calc Zspec and MTRasym
M_z = load(['M_z_APTw_3T_000_2uT_1block_2s_braintumor.seq.txt']);
M_zref = M_z(end:-1:1);
MTRasym = M_zref(1:end-1)-M_z(2:end);


%% 3)  Plots and further graphics

% Zspec and MTrasym 
figure;
subplot(1,2,1)
plot(defs.offsets_ppm(2:end),M_z(2:end),'r.-')
subplot(1,2,2)
plot(defs.offsets_ppm(2:end),MTRasym,'b.-')

% Parametric Plot of MTRasym at 3.5 ppm
figure, imagesc(squeeze(V_MTRasym(31,:,:,6)),[-0.05 0.05]);


%=================================================================================================================================



%% PULSEQ-CEST MEASURMENT
%=================================================================================================================================
%% 1)  read in data from measurement (dicom)
opath=uigetdir('','Go to DICOM Directory');cd(opath)

% Question if seq File is still the same
question = input('Are the DICOM Files acquired with the same Protocoll Parameters from the PulseqCEST Library? [y/n]','s')
if strcmpi(question, 'y')
else
[seqfile vpath]=uigetfile('','.seq');cd(vpath)
seq.read(seqfile);
defs.offsets_ppm   = seq.definitions('offsets_ppm');
Nmeas=numel(defs.offsets_ppm);
end


cd(opath)
collection = dicomCollection(opath);
V = dicomreadVolume(collection); sz=size(V); V=reshape(V,[sz(1) sz(2) Nmeas sz(4)/Nmeas ]); V= permute(V,[1 2 4 3]); size(V)
%figure;subplot(1,2,1), imagesc(V(:,:,6,1));  subplot(1,2,2), plot(squeeze(V(50,50,1,:)));

%% 2)  Vectorization Forwards
sz=size(V);
maskInd=1:sz(1)*sz(2)*sz(3);
V_M_z=permute(V,[4 1 2 3]); M_z=V_M_z(:,maskInd); M_z=double(M_z); 
%figure, plot(M_z(:,5050:5060));
%figure, montage1t(squeeze(V_M_z(4,:,:,:)))

%% 3)  B0 Correction
M0=M_z(1,:);
Z=M_z(2:end,:)./M0; 
w=defs.offsets_ppm(2:end);
Z_corr=zeros(size(Z,1),size(Z,2)); dB0_stack=zeros(1,size(Z,2));
% Perform smoothing spline interpolation
tic
for ii=1:size(Z,2)
    try
        if  isfinite(Z(:,ii))
            pp = csaps(w, Z(:,ii), 0.95);
            w_fine=-1:0.005:1;
            Z_fine = ppval(pp, w_fine); 
            [~, MINidx] = min(Z_fine);
            dB0=w_fine(MINidx);
            dB0_stack(1,ii)=dB0;
            Z_corr(:,ii)= ppval(pp, w+dB0); 
    %         disp(ii/size(Z,2));
        end
    end
end
toc

%% 4)  Calc Zspec and MTRasym
% [Z_corrExt_denoised,used_components] = pcca_3D(Z_corrExt,P, (1:numel(P.SEQ.w)), 0, Segment); %principle component analysis (for denoising)
Zref = Z_corr(end:-1:1,:);
MTRasym = Zref-Z_corr;

%Vectorization Backwards
if size(Z,2)>0
    V_MTRasym=double(V_M_z(2:end,:,:,:))*0;   
    V_MTRasym(:,maskInd)=MTRasym;
    V_Z_corr=double(V_M_z(2:end,:,:,:))*0; 
    V_Z_corr(:,maskInd)=Z_corr; 
end


%% 5)  Plots and further graphics
sliceofinterest=6;                   % Pick slice for Evaluation

figure;
subplot(1,3,1);imshow(squeeze(V_Z_corr(sliceofinterest,:,:,1)),[0.5 1]);[aa bb]=ginput(1);aa=round(aa);bb=round(bb);hold on;plot(aa,bb,'r+');
subplot(1,3,2);plot(w,V_Z_corr(:,bb,aa,sliceofinterest),'r.-');
subplot(1,3,3);plot(w,V_MTRasym(:,bb,aa,sliceofinterest),'b.-')

offsetofinterest=find(defs.offsets_ppm == 3.5);
figure, imagesc(squeeze(V_MTRasym(offsetofinterest,:,:,sliceofinterest)),[-0.05 0.05]);
%colormap(gca,RAINBOW)
%=================================================================================================================================

File: seq-library/APTw_3T_000_2uT_1block_2s_braintumor/EVAL_APTw_3T_000_2uT_1block_2s_braintumor.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun 28 13:12:35 2023

@author: kouemoin
"""

import numpy as np
import matplotlib.pyplot as plt
import pypulseq as pp
from csaps import csaps
import os
import  pydicom
seq = pp.Sequence()


seq_path = 'W:/radiologie/mr-physik-data/Mitarbeiter/kouemo/matlab/pulseq-cest-library/seq-library/APTw_3T_000_2uT_1block_2s_braintumor/APTw_3T_000_2uT_1block_2s_braintumor.seq'  # can be a str or a Path

seq.read(seq_path)

m0_offset = seq.get_definition('M0_offset')

offsets = seq.get_definition('offsets_ppm')

Nmeas = len(offsets)

#change to the directory containing the Dicom files
dcmpath = 'W:/radiologie/mr-physik-data/Mitarbeiter/kouemo/Testordner/dcm/PULSEQ_HYBRID_GRE_2_2_5_APTW_001_RR_0015'
os.chdir(dcmpath)

#read data from dicom directory
collection = [pydicom.dcmread(filename) for filename in os.listdir(dcmpath)]

#extract the volume data
V = np.stack([dcm.pixel_array for dcm in collection])
V = np.transpose(V, (1, 2, 0))
sz = V.shape
V = np.reshape(V, [sz[0], sz[1], Nmeas, sz[2] // Nmeas]).transpose(0,1,3,2)

#Vectorization
mask = np.squeeze(V[:,:,:,0]) > 100
mask_idx = np.where(mask.ravel())[0]
V_m_z =  V.reshape(-1, Nmeas).T
m_z = V_m_z[:, mask_idx]

M0_idx = np.where(offsets == m0_offset)[0]

if len(M0_idx) > 0:
    M0 = np.mean(m_z[M0_idx,:],0)
    offsets = np.delete(offsets, M0_idx)
    m_z = np.delete(m_z, M0_idx,axis=0)
    Z = m_z / M0 #Normalization
else:
    print('m0_offset not found in offset')


#helper function to evaluate piecewise polinomial 
def ppval(p, x):
    if callable(p):
        return p(x)
    else:
        n = len(p) - 1
        result = np.zeros_like(x)
        for i in range(n, -1, -1):
            result = result * x + p[i]
        return result

#perform the smoothing spline interpolation
Z_corr = np.zeros_like(Z)
w = offsets
dB0_stack = np.zeros(Z.shape[1])
for ii in range(Z.shape[1]):
     if np.all(np.isfinite(Z[:, ii])):
         pp = csaps(w, Z[:,ii], smooth=0.95)
         w_fine = np.arange(-1, 1.005, 0.005)
         z_fine = ppval(pp, w_fine)
            
         min_idx = np.argmin(z_fine)
         dB0_stack[ii] = w_fine[min_idx]
            
         Z_corr[:,ii] = ppval(pp, w + dB0_stack[ii])
    
    
# #calc of MTRasym-Spectrum
Z_ref = Z_corr[::-1,:]
MTRasym = Z_ref - Z_corr

#Vectorization Backwards
if Z.shape[1] > 1:
    
    V_MTRasym = np.zeros((V_m_z.shape[0], V_m_z.shape[1]), dtype=float)
    V_MTRasym[1:, mask_idx] = MTRasym
    V_MTRasym_reshaped = V_MTRasym.reshape(V.shape[3], V.shape[0], V.shape[1], V.shape[2]).transpose(1,2,3,0)
    
    V_Z_corr = np.zeros((V_m_z.shape[0], V_m_z.shape[1]), dtype=float)
    V_Z_corr[1:, mask_idx] = Z_corr
    V_Z_corr_reshaped = V_Z_corr.reshape(V.shape[3], V.shape[0], V.shape[1], V.shape[2]).transpose(1,2,3,0)
    
#plots and further graphics
plt.figure(figsize=(10, 4))
plt.subplot(1, 2, 1)
plt.plot(w, np.mean(Z_corr, axis=1), 'r.-')
plt.gca().invert_xaxis()
plt.title('Mean Z-spectrum')

plt.subplot(1, 2, 2)
plt.plot(w, np.mean(MTRasym, axis=1), 'b.-')
plt.xlim([0, 4])
plt.gca().invert_xaxis()
plt.title('Mean MTRasym-spectrum')
plt.show()


  
 


    
    
    